### PR TITLE
Flutter iOS Interactive Keyboard: Take Screenshot and Handle Pointer Movement 

### PR DESCRIPTION
**Pull request by Matt2D**
This PR address the movement aspect of the flutter interactive keyboard. It handles pointer movement while a scroll view widget is visible, and the interactive behavior is chosen for keyboardDismissBehavior. This is a desired behavior of the keyboard that has not yet been implemented. 
Design Document:
https://docs.google.com/document/d/1-T7_0mSkXzPaWxveeypIzzzAdyo-EEuP5V84161foL4/edit?pli=1

Issues Address:
https://github.com/flutter/flutter/issues/57609

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.h
Comment (hellohuanlin) [inline shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.h:169]: Can you rename to `flt_firstResponder` in case of naming conflict? (ObjC doesn't support namespace)

File: shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm
Comment (hellohuanlin) [inline shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm:48]: Nit maybe `kOnInteractiveKeyboardPointerMoveMethod = @"TextInput.onPointerMoveForInteractiveKeyboard"`, to be more specific.

File: shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm
Comment (hellohuanlin) [inline shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm:2293]: use `CGRect` rather than `double` for UI stuff.

**Review by hellohuanlin — COMMENTED** [review]
Overall looks great. I will ask @cbracken to also take a look.

File: shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm
Comment (Matt2D) [reply shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm:2294]: With regards to the pointerY this only contains a single double, so I am not sure that this would classify as UI.

Comment (hellohuanlin) [reply shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm:2294]: sorry i meant CGFloat

File: shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm
Comment (cyanglaz) [inline shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm:2305]: nits:
```suggestion
  // Determines if pointer goes overtop of the keyboard.
```

File: shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm
Comment (cyanglaz) [inline shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm:2308]: Using `_keyboardViewContainer.subviews.count == 0` doesn't seems to be safe and future proof. For example, what if someone added a new subview for `_keyboardViewContainer ` and didn't know this logic exist? Could you use a different indicator?

File: shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm
Comment (cyanglaz) [inline shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm:2314]: nits: could you refactor these lines into a method to match what you did in `if`
Maybe something like:
`[self setKeyboardContainerHeight:pointerY]

File: shell/platform/darwin/ios/framework/Source/FlutterTextInputPluginTest.mm
Comment (cyanglaz) [inline shell/platform/darwin/ios/framework/Source/FlutterTextInputPluginTest.mm:63]:
```suggestion
@property(nonatomic, weak) UIView* keyboardViewContainer;
```

**Review by cyanglaz — CHANGES_REQUESTED** [review]

File: shell/platform/darwin/ios/framework/Source/FlutterTextInputPluginTest.mm
Comment (hellohuanlin) [reply shell/platform/darwin/ios/framework/Source/FlutterTextInputPluginTest.mm:63]: @cyanglaz i could be wrong, but iirc the standard practice is just `assign` when exposing the property in test header (similar to the `activeView` properly above). `weak` is confusing since it's actually a `strong` property.

Comment (cyanglaz) [reply shell/platform/darwin/ios/framework/Source/FlutterTextInputPluginTest.mm:63]: Oh. I didn't realize it is only exposing the property. I'm not aware of the standard practice tho but since activeView also used assign, I will take your word for it :)

Comment (hellohuanlin) [reply shell/platform/darwin/ios/framework/Source/FlutterTextInputPluginTest.mm:63]: @cyanglaz and I just discussed offline - it should be `readonly` here, since your unit test does not need the setter. 

```
@property(nonatomic, readonly) UIView* keyboardViewContainer;
```

File: shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.h
Comment (Matt2D) [reply shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.h]: Good idea! Went through and made the change.

File: shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm
Comment (Matt2D) [reply shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm]: Just updated, I'll make sure to try and add Interactive Keyboard more in the future, definitely would help with clarity!

File: shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm
Comment (Matt2D) [reply shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm]: Gotcha, just updated the code!

File: shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm
Comment (Matt2D) [reply shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm]: Will do! Just slipped by me.

File: shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm
Comment (Matt2D) [reply shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm]: Definitely a concern, the same thing occurs in the tests so there definitely is a need to update this approach. I went through and included a UIView that represents the screenshot. Then we can see if it the superview matches instead.

File: shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm
Comment (Matt2D) [reply shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm]: Done! Always helpful to simplify code into sections to increase understandability.

File: shell/platform/darwin/ios/framework/Source/FlutterTextInputPluginTest.mm
Comment (Matt2D) [reply shell/platform/darwin/ios/framework/Source/FlutterTextInputPluginTest.mm]: Gotcha! That works for me, just went through and made the changes.

File: shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm
Comment (cyanglaz) [reply shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm:2304]: Is the pointerY based on the screen, or the flutter app?

Since it comes from the Framework, I assume it is based on the flutter app? So In scenarios (add to app, or app extension) when the flutter app does not fill the full screen, the calculation below that determines if the pointer goes overtop of the keyboard should be based on the flutter app's height instead of the screen height, correct?

Maybe instead of the screenHeight, we should use the height of the FlutterViewController's view?

File: shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm
Comment (cyanglaz) [inline shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm:2309]: Did you mean `if (_keyboardView.superview == nil)`

File: shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm
Comment (mossmana) [reply shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm:2304]: > We are moving away from using [UIScreen mainScreen]. I don't think we should block this PR because of this, but we will need to follow up to clean up later. FYI @mossmana

@Matt2D While usually I would agree with not blocking the PR, in this particular case, this is a new introduction of [UIScreen mainScreen]. I would prefer that we don't add new instances. You can refer to https://github.com/flutter/flutter/issues/125496 for how we are handling the deprecation in other places.

Comment (hellohuanlin) [reply shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm:2304]: @cyanglaz and I were discussing this the other day, and it's actually a bit tricky here - we would need to refactor [this code](https://github.com/flutter/engine/pull/43690/files#diff-365b13daa5168d36d5824f220bc472a4a90f4685761947f8daf9dc76e7f8d962R480-R497) to make it re-usable. 

I think I can take this part over. Can make a PR today.

File: shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.h
Comment (cbracken) [reply shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.h:169]: Avoid underscore in obj-c identifiers. We're doing something similar in another outstanding PR on the macOS side and using `flutter` as the prefix here:
https://github.com/flutter/engine/pull/43101/files#diff-edbbe0be10d99338c2496cad85a27776c79ec8b260825ee4d309b1e4d194d5b2R10

So we'd end up with `flutterFirstResponder` here.

@cyanglaz @hellohuanlin preferences? We should probably be consistent across embedders.

File: shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm
Comment (cbracken) [reply shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm:2304]: @cyanglaz in this case, since this is being used to determine the pointer position on the screen, this is relative to the screen rather than the flutter view.

File: shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm
Comment (cbracken) [inline shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm:2306]: Maybe something like:
```suggestion
  // If the pointer is within the bounds of the keyboard.
```

Comment (cbracken) [reply shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm:2306]: (And maybe move the comment inside the conditional block)

File: shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm
Comment (cbracken) [inline shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm:2336]: Consider renaming to something more specific for readability -- e.g. `take KeyboardSnapshot`

File: shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm
Comment (cbracken) [reply shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm:2345]: Since this is positioning a keyboard relative to the full screen -- regardless of whether the FlutterView is full screen, or a subview covering only part of the screen in an add-to-app scenario -- I think we want the root view controller don't we?

File: shell/platform/darwin/ios/framework/Source/FlutterTextInputPluginTest.mm
Comment (cbracken) [inline shell/platform/darwin/ios/framework/Source/FlutterTextInputPluginTest.mm:2404]: Consider just inlining this into the call below.

File: shell/platform/darwin/ios/framework/Source/FlutterTextInputPluginTest.mm
Comment (cbracken) [inline shell/platform/darwin/ios/framework/Source/FlutterTextInputPluginTest.mm:2441]: Similar comment here.

File: shell/platform/darwin/ios/framework/Source/FlutterTextInputPluginTest.mm
Comment (cbracken) [inline shell/platform/darwin/ios/framework/Source/FlutterTextInputPluginTest.mm:2412]: Maybe rather than just "up" something like "back to original position"?

**Review by cbracken — COMMENTED** [review]
Overall looks great - just a few comments aside from those left by others!

File: shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm
Comment (cyanglaz) [reply shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm:2345]: Ah yeah that's true. We would need to make sure to position the keyboard screenshot on the very top. Set ZIndex to maxInt maybe?

File: shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.h
Comment (cyanglaz) [reply shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.h:169]: `flutterFirstResponder` SGTM

Comment (Matt2D) [reply shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.h:169]: Will do! Thank you for the suggestion

File: shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm
Comment (Matt2D) [reply shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm:2306]: Makes sense, thank you!

File: shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm
Comment (hellohuanlin) [reply shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm:2304]: FYI this is the PR promised: https://github.com/flutter/engine/pull/44303

After that gets landed, you can do something like this here: 

```
UIScreen *screen = _viewController.flutterScreenIfViewLoaded;
if (!screen) {
  return;
}
CGFloat screenHeight = screen.bounds.size.height;
// ...
```

File: shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm
Comment (Matt2D) [reply shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm:2326]: This works great! The one change I added is that we still want to check if the pointer moves over the keyboard before we take the screenshot.

File: shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm
Comment (Matt2D) [reply shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm:2336]: Definitely, I also added add to screen to be even better for readability.

File: shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm
Comment (Matt2D) [reply shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm:2337]: Updated with the fix provided by the new PR Huan did.

File: shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm
Comment (Matt2D) [reply shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm:2345]: Definitely worth adding!

File: shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm
Comment (Matt2D) [reply shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm]: Thank you for the code snippet, I think the new api is much cleaner! Updated the code using the mentioned solution.

File: shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm
Comment (Matt2D) [reply shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm]: Yep you are right! Thanks for catching this.

File: shell/platform/darwin/ios/framework/Source/FlutterTextInputPluginTest.mm
Comment (Matt2D) [reply shell/platform/darwin/ios/framework/Source/FlutterTextInputPluginTest.mm]: Definitely!

File: shell/platform/darwin/ios/framework/Source/FlutterTextInputPluginTest.mm
Comment (Matt2D) [reply shell/platform/darwin/ios/framework/Source/FlutterTextInputPluginTest.mm]: Done

File: shell/platform/darwin/ios/framework/Source/FlutterTextInputPluginTest.mm
Comment (Matt2D) [reply shell/platform/darwin/ios/framework/Source/FlutterTextInputPluginTest.mm]: Good idea! Done

File: shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm
Comment (cyanglaz) [reply shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm:2352]: Ah I missed that part. So the caller part can probably be
```
if (_keyboardView.superview != nil) {
   [self setKeyboardContainerHeight:pointerY];
} else if (screenHeight - keyboardHeight <= pointerY) {
   [self takeScreenshot];
   [self hideKeyboardWithoutAnimation];
}
```

**Review by cyanglaz — APPROVED** [review]
LGTM

**Review by hellohuanlin — APPROVED** [review]

**Comment by hellohuanlin** [comment]
You got a build error in test: 

```
../../flutter/shell/platform/darwin/ios/framework/Source/FlutterTextInputPluginTest.mm:2515:72: error: C-style cast from 'NSNumber *' to 'CGFloat' (aka 'double') is not allowed
 2515 |   XCTAssertEqual(textInputPlugin.keyboardViewContainer.frame.origin.y, (CGFloat) @(600));
```

You may have to do 
```
XCTAssertEqual(textInputPlugin.keyboardViewContainer.frame.origin.y, 600.0);
```

File: shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm
Comment (hellohuanlin) [reply shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm:2352]: Discussed offline - the duplicate logic of `screenHeight - keyboardHeight <= pointerY` can get error prune. 

Another idea is to pass in an additional BOOL to the helper function, but that also makes the code unnecessarily complicate. 

I think it makes sense to keep it as it is.

**Pull request merged**
